gpu/src/cuda/abstractions/state.rs

```diff
@@ -0,0 +1,98 @@
+use crate::cuda::abstractions::{element::CUDAFieldElement, errors::CudaError};
+use cudarc::{
+    driver::{CudaDevice, LaunchAsync, LaunchConfig},
+    nvrtc::safe::Ptx,
+};
+
+const FFT_PTX: &str = include_str!("../shaders/fft.ptx");
+
+/// Structure for abstracting basic calls to a Metal device and saving the state. Used for
+/// implementing GPU parallel computations in Apple machines.
+pub(crate) struct CudaState {
+    pub(crate) device: CudaDevice,
+    pub(crate) library: metal::Library,
+    pub(crate) queue: metal::CommandQueue,
+}
+
+impl CudaState {
+    /// Creates a new CUDA state with the first GPU.
+    pub(crate) fn new() -> Result<Self, CudaError> {
+        let device =
+            CudaDevice::new(0).map_err(|err| CudaError::DeviceNotFound(err.to_string()))?;
+
+        // Load PTX libraries
+        // TODO: extract into helper method
+        device
+            .load_ptx(
+                Ptx::from_src(/* Library code */ FFT_PTX),
+                /* Library name */ "fft",
+                /* defined functions */ &["radix2_dit_butterfly"],
+            )
+            .map_err(|err| CudaError::PtxError(err.to_string()))?;
```

Comment:
Will we load all modules (`fft`, `gen_twiddles` and `bitrev_permutation`) at new()?

Reply:
I was thinking of moving the module loading to the corresponding `get_<function>` implementations, but I think we should benchmark it first.

Reply:
For now don't think too much about these kinds of optimizations, I think PTX loading is pretty small compared to what FFT takes to run